File: web/src/algorithms/sorting/counting-sort.ts
import type { AlgorithmImplementation, AlgorithmStep } from '@/lib/types/algorithm';

export const countingSort: AlgorithmImplementation = {
  id: 'counting-sort',
  name: 'Counting Sort',
  category: 'sorting',
  timeComplexity: { best: 'O(n + k)', average: 'O(n + k)', worst: 'O(n + k)' },
  spaceComplexity: 'O(n + k)',
  pseudocode: [
    { line: 0, text: 'procedure countingSort(A)' },
    { line: 1, text: '  k ← max(A)' },
    { line: 2, text: '  create count[0..k] = 0' },
    { line: 3, text: '  for each element x in A do' },
    { line: 4, text: '    count[x] ← count[x] + 1' },
    { line: 5, text: '  for i ← 1 to k do' },
    { line: 6, text: '    count[i] ← count[i] + count[i-1]' },
    { line: 7, text: '  for j ← n-1 downto 0 do' },
    { line: 8, text: '    output[count[A[j]] - 1] ← A[j]' },
    { line: 9, text: '    count[A[j]] ← count[A[j]] - 1' },
    { line: 10, text: '  return output' },
  ],

  generateSteps(input: number[]): AlgorithmStep[] {
    const steps: AlgorithmStep[] = [];
    const n = input.length;

    if (n === 0) return steps;

    // Normalize: ensure all values are non-negative
    const minVal = Math.min(...input);
    const array = input.map(v => v - minVal);
    const k = Math.max(...array);

    // Phase 1: Count occurrences
    const count: number[] = new Array(k + 1).fill(0);

    steps.push({
      type: 'highlight',
      array: [...array],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [],
      pseudocodeLine: 2,
      description: `Initialize count array of size ${k + 1}. Values range: ${minVal} to ${minVal + k}`,
    });

    // Count phase
    for (let i = 0; i < n; i++) {
      count[array[i]]++;

      steps.push({
        type: 'select',
        array: [...count],
        highlightedIndices: [array[i]],
        secondaryIndices: [],
        sortedIndices: [],
        pseudocodeLine: 4,
        description: `Count element ${array[i] + minVal}: count[${array[i]}] = ${count[array[i]]}`,
      });
    }

    steps.push({
      type: 'pass-complete',
      array: [...count],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [],
      pseudocodeLine: 4,
      description: `Count phase complete: [${count.join(', ')}]`,
    });

    // Cumulate phase
    for (let i = 1; i <= k; i++) {
      steps.push({
        type: 'compare',
        array: [...count],
        highlightedIndices: [i, i - 1],
        secondaryIndices: [],
        sortedIndices: [],
        pseudocodeLine: 5,
        description: `Cumulate: count[${i}] = ${count[i]} + count[${i - 1}] = ${count[i - 1]}`,
      });

      count[i] += count[i - 1];

      steps.push({
        type: 'insert',
        array: [...count],
        highlightedIndices: [i],
        secondaryIndices: [],
        sortedIndices: [],
        pseudocodeLine: 6,
        description: `count[${i}] = ${count[i]}`,
      });
    }

    steps.push({
      type: 'pass-complete',
      array: [...count],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [],
      pseudocodeLine: 6,
      description: `Cumulative count: [${count.join(', ')}]`,
    });

    // Place phase: build output array
    const output: number[] = new Array(n).fill(0);
    const sortedIndices: number[] = [];

    for (let j = n - 1; j >= 0; j--) {
      const val = array[j];
      const pos = count[val] - 1;
      output[pos] = val + minVal;
      count[val]--;
      sortedIndices.push(pos);

      steps.push({
        type: 'traverse',
        array: [...output],
        highlightedIndices: [pos],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 8,
        description: `Place ${val + minVal} (from input[${j}]) at output[${pos}]`,
      });
    }

    // Final sorted array
    steps.push({
      type: 'sorted',
      array: [...output],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: Array.from({ length: n }, (_, i) => i),
      pseudocodeLine: 10,
      description: 'Counting Sort complete. Array is sorted.',
    });

    return steps;
  },
};

File: apple/AestheticAlgorithm/Algorithms/Sorting/CountingSort.swift
import Foundation

struct CountingSort: AlgorithmExecutable {
    let id = "counting-sort"
    let name = "Counting Sort"
    let category = "sorting"
    let timeComplexity = (best: "O(n + k)", average: "O(n + k)", worst: "O(n + k)")
    let spaceComplexity = "O(n + k)"

    let pseudocode: [(line: Int, text: String)] = [
        (0, "procedure countingSort(A)"),
        (1, "  k ← max(A)"),
        (2, "  count[0..k] ← 0"),
        (3, "  // Count phase"),
        (4, "  for each element in A do"),
        (5, "    count[element] ← count[element] + 1"),
        (6, "  // Cumulative count"),
        (7, "  for i ← 1 to k do"),
        (8, "    count[i] ← count[i] + count[i-1]"),
        (9, "  // Placement phase"),
        (10, "  for i ← n-1 downto 0 do"),
        (11, "    output[count[A[i]] - 1] ← A[i]"),
        (12, "    count[A[i]] ← count[A[i]] - 1"),
        (13, "  return output")
    ]

    func generateSteps(from input: [Int]) -> [AlgorithmStep] {
        var steps: [AlgorithmStep] = []
        let array = input

        if array.isEmpty { return steps }

        let n = array.count
        let maxVal = array.max() ?? 0
        let k = maxVal + 1

        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 0,
            description: "Starting Counting Sort on \(n) elements. Max value = \(maxVal)"
        ))

        // Initialize count array
        var count = [Int](repeating: 0, count: k)

        steps.append(AlgorithmStep(
            type: .highlight,
            array: count,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 2,
            description: "Initialized count array of size \(k) to all zeros"
        ))

        // Count phase
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 3,
            description: "Phase 1: Counting occurrences of each element"
        ))

        for i in 0..<n {
            let val = array[i]

            // Highlight element being counted
            steps.append(AlgorithmStep(
                type: .select,
                array: array,
                highlightedIndices: [i],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 4,
                description: "Counting element \(val) at index \(i)"
            ))

            count[val] += 1

            // Show updated count array
            steps.append(AlgorithmStep(
                type: .insert,
                array: count,
                highlightedIndices: [val],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 5,
                description: "count[\(val)] = \(count[val])"
            ))
        }

        // Cumulative count phase
        steps.append(AlgorithmStep(
            type: .highlight,
            array: count,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 6,
            description: "Phase 2: Computing cumulative counts"
        ))

        for i in 1..<k {
            steps.append(AlgorithmStep(
                type: .compare,
                array: count,
                highlightedIndices: [i],
                secondaryIndices: [i - 1],
                sortedIndices: [],
                pseudocodeLine: 8,
                description: "count[\(i)] = count[\(i)] + count[\(i-1)] = \(count[i]) + \(count[i-1])"
            ))

            count[i] += count[i - 1]

            steps.append(AlgorithmStep(
                type: .insert,
                array: count,
                highlightedIndices: [i],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 8,
                description: "count[\(i)] = \(count[i])"
            ))
        }

        // Placement phase
        var output = [Int](repeating: 0, count: n)
        var placedIndices: [Int] = []

        steps.append(AlgorithmStep(
            type: .highlight,
            array: output,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 9,
            description: "Phase 3: Placing elements into sorted output"
        ))

        for i in stride(from: n - 1, through: 0, by: -1) {
            let val = array[i]

            // Show element being placed
            steps.append(AlgorithmStep(
                type: .select,
                array: array,
                highlightedIndices: [i],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 10,
                description: "Placing element \(val) from input[\(i)]"
            ))

            count[val] -= 1
            let outputIndex = count[val]
            output[outputIndex] = val
            placedIndices.append(outputIndex)

            // Show element placed in output
            steps.append(AlgorithmStep(
                type: .insert,
                array: output,
                highlightedIndices: [outputIndex],
                secondaryIndices: [],
                sortedIndices: placedIndices.sorted(),
                pseudocodeLine: 11,
                description: "output[\(outputIndex)] = \(val). count[\(val)] decremented to \(count[val])"
            ))
        }

        // Final sorted output
        steps.append(AlgorithmStep(
            type: .sorted,
            array: output,
            highlightedIndices: Array(0..<n),
            secondaryIndices: [],
            sortedIndices: Array(0..<n),
            pseudocodeLine: 13,
            description: "Counting Sort complete"
        ))

        return steps
    }
}
